### PR TITLE
[doc][api/6] convert api alias to full name

### PR DESCRIPTION
**Pull request by can-anyscale**
- Add _fullname function to convert an api alias to its full name. For example, in our doc, `ray.data.range` is and alias and used instead of its full name `ray.data.read_api.range`. These alias is defined in `ray/data/__init__.py`
- This is useful so that the analysis can use the same canonical name between apis declared in different sources

Test:
- created alias for the mock_module; test that the api always have a the full name
- CI

### CODE REVIEW

File: ci/ray_ci/doc/api.py [hunk 1]
```diff
@@ -27,6 +28,7 @@ class API:
     annotation_type: AnnotationType
     code_type: CodeType
 
+    @staticmethod
```
Comment (can-anyscale) [inline ci/ray_ci/doc/api.py:31]: add here just for completeness, not logic change

File: ci/ray_ci/doc/mock/__init__.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+from ci.ray_ci.doc.mock.mock_module import MockClass
+from ci.ray_ci.doc.mock.mock_module import mock_function
+
+__all__ = [
+    "MockClass",
```
Comment (can-anyscale) [inline ci/ray_ci/doc/mock/__init__.py:5]: this creates `ci.ray_ci.doc.mock.MockClass` as an alias for the full name `ci.ray_ci.doc.mock.mock_module.MockClass`

File: ci/ray_ci/doc/test_api.py [hunk 3]
```diff
@@ -10,7 +11,9 @@
 )
 
 
-def test_from_autosummary():
+@mock.patch("ci.ray_ci.doc.api.API._fullname")
```
Comment (can-anyscale) [inline ci/ray_ci/doc/test_api.py:14]: mocking this to keep all data in this test file as fake; i tested the e2e integration of this function with autodoc in the `test_autodoc`

File: ci/ray_ci/doc/test_autodoc.py [hunk 4]
```diff
@@ -14,29 +14,24 @@ def test_walk():
...
         with open(os.path.join(tmp, "api_01.rst"), "w") as f:
+            f.write(".. currentmodule:: ci.ray_ci.doc.mock\n")
             f.write(".. autosummary::\n\n")
-            f.write("\tfunc_01\n")
-            f.write("\tfunc_02\n")
+            f.write("\tmock_function\n")
         with open(os.path.join(tmp, "api_02.rst"), "w") as f:
-            f.write(".. currentmodule:: mymodule\n")
-            f.write(".. autoclass:: class_01\n")
+            f.write(".. currentmodule:: ci.ray_ci.doc.mock\n")
```
Comment (can-anyscale) [inline ci/ray_ci/doc/test_autodoc.py:21]: use real name here for integration testing

File: ci/ray_ci/doc/BUILD.bazel [hunk 5]
```diff
@@ -23,7 +24,7 @@ py_library(
 py_test(
     name = "test_module",
     size = "small",
-    srcs = ["test_module.py", "mock_module.py"],
+    srcs = ["test_module.py", "__init__.py", "mock/mock_module.py", "mock/__init__.py"],
```
Comment (aslonnie) [inline ci/ray_ci/doc/BUILD.bazel:27]: maybe split the mock / test file into a separate sub directory? it is quite confusing that `__init__.py` is a test only file here.

File: ci/ray_ci/doc/api.py [hunk 6]
```diff
@@ -83,10 +86,28 @@ def from_autoclass(
...
+
+    @staticmethod
+    def _fullname(name: str) -> str:
+        """
+        Some APIs have aliases declared in __init__.py file (see ray/data/__init__.py
+        for example). This method converts the alias to full name. This is to make sure
+        out analysis can be performed on the same set of canonial names.
+        """
+        tokens = name.split(".")
+        attribute = importlib.import_module(tokens[0])
```
Comment (aslonnie) [inline ci/ray_ci/doc/api.py:106]: this is just to import the first token, like as a package? I am confused how this works.

could you give an example (or a unit test) of how this function works?

Comment (can-anyscale) [reply ci/ray_ci/doc/api.py:108]: i added a test case in `test_api.py`; the high level idea is given the python path as string, e.g. a.b.c; this logic convert the string into a python object, by importing each token one by one ... import a, then import a.b, then import a.b.c, etc; once the python object is obtained, obtain its full name

File: ci/ray_ci/doc/test_api.py [hunk 7]
```diff
@@ -100,5 +106,12 @@ def test_from_autoclasss():
         ) == str(test["output"])
 
 
+def test_fullname():
+    assert (
+        API._fullname("ci.ray_ci.doc.mock.mock_function")
```
Comment (can-anyscale) [inline ci/ray_ci/doc/test_api.py:111]: test case

**Comment by can-anyscale** [comment]
@aslonnie's comments; move all __init__ into mock directory

File: ci/ray_ci/doc/api.py [hunk 8]
```diff
@@ -83,10 +86,30 @@ def from_autoclass(
...
             annotation_type=AnnotationType.PUBLIC_API,
             code_type=CodeType.CLASS,
         )
+
+    @staticmethod
+    def _fullname(name: str) -> str:
+        """
+        Some APIs have aliases declared in __init__.py file (see ray/data/__init__.py
+        for example). This method converts the alias to full name. This is to make sure
+        out analysis can be performed on the same set of canonial names.
```
Comment (aslonnie) [inline ci/ray_ci/doc/api.py:103]: nit: out -> our

File: ci/ray_ci/doc/api.py (hunk 6 above)
Comment (aslonnie) [reply ci/ray_ci/doc/api.py:108]: I am not sure if this is doing the right thing..

like:

````
>>> print(ray.init.__module__)
ray._private.worker
````

where the symbol is implemented does not always mean where the symbol is exposed / belongs to.

Comment (aslonnie) [reply ci/ray_ci/doc/api.py:108]: maybe you can use this as a signature to determine that two symbols from different packages are the same one, but I am not sure that the alias is just a shortcut, or it is an exposed and even high-ranked interface.

**Comment by can-anyscale** [comment]
chat offline, the concern is that we're replacing the api.name with _fullname; i now keep the field name as it is, and replace _fullname as a function for the API class, so that someone can use it to find the root of an API

**Review by aslonnie — APPROVED** [review]

**Pull request merged**